File: src/main/scala/clase6/ADTUnbundledStatefulSecure.scala
package clase6

object ADTUnbundledStatefulSecure extends App {

  private case class Key()

  case class Stack[T](private var content:List[T]) {
    def unwrap(k:Key):List[T] = content
    def wrap(k:Key, content:List[T]):Unit = {
      this.content = content
    }
  }

  def stack[T](): Stack[T] = Stack[T](Nil)

  def push[T](s: Stack[T], element: T) = {
    s.wrap(Key(), element :: s.unwrap(Key()))
  }

  def pop[T](s: Stack[T]) = {
    s.unwrap(Key()) match {
      case head :: tail =>
        s.wrap(Key(), tail)
        head
    }
  }

  def isEmpty[T](s: Stack[T]) = s.unwrap(Key()) == Nil

  assert(isEmpty(stack()))

  val s = stack[String]()

  push(s, "ariel")
  push(s, "alejandro")

  assert(!isEmpty(s))

  assert(pop(s) == "alejandro")
  assert(pop(s) == "ariel")
  assert(isEmpty(s))

}

File: haskell/src/clase6/ADTUnbundledStatefulSecure.hs
module ADTUnbundledStatefulSecure where

import Debug.Trace
import Data.IORef

-- private type
data Key = Key

data Wrapper t = Wrapper {
  -- data is actually stored in the closure's environment
  content :: Key -> [t]
}

wrap value = Wrapper (\k -> value)

unwrap wrapper key = ((content wrapper) key)

stack = newIORef (wrap [])

push stack element = modifyIORef stack (\s -> wrap (element : (unwrap s Key)))

pop stack = do
  current <- readIORef stack
  case (unwrap current Key) of
    (head : tail) -> do
      writeIORef stack (wrap tail)
      return head

isEmpty stack = do
  current <- readIORef stack
  return (case (unwrap current Key) of
    [] -> True
    _ -> False)

main = do
  s <- stack
  isEmpty s >>= traceShowM -- True
  push s "ariel"
  push s "alejandro"
  isEmpty s >>= traceShowM -- False
  pop s >>= traceShowM -- "alejandro"
  pop s >>= traceShowM -- "ariel"
  isEmpty s >>= traceShowM -- True
g